tests/test_tasks.py

```diff
@@ -640,6 +640,28 @@ def test_flow():
             assert y == i + 1
             assert exceptions_equal(z, ValueError("Fail task"))
 
+    async def test_raises_if_depends_on_itself(self):
+        @task
+        def say_hello(name):
+            return f"Hello {name}!"
+
+        @flow
+        def my_flow():
+            greeting_queue = []
+            for i in range(3):
+                if greeting_queue:
+                    wait_for = greeting_queue
+                else:
+                    wait_for = []
+                future = say_hello.submit(name=f"Person {i}", wait_for=wait_for)
+                greeting_queue.append(future)
```

Comment:
Hmm, so the only way this can happen is if the caller changes the `wait_for` iterable after submitting the task.  If we just make sure to make our own copy of `wait_for`, doesn't this problem go away?

Reply:
@desertaxle pointed out copying could cause memory issues. Maybe better to start simple and copy if we need to later?

Reply:
It wouldn't be a copy of the futures, just a copy of the iterable the caller gave us so that it's kind of "causally locked in" at that point (I'm just thinking literally `wait_for = list(wait_for)`.  I can't help but think there are other classes of bugs lurking if we let this list of futures be mutable after it's passed to the engine.

Reply:
This also happens if futures are passed in parameters which can be deeply nested and would require a `deepcopy` to prevent subsequent modification from causing deadlocks. This probably won't happen often, so I'd rather raise in the unlikely chance this happens then incur a memory penalty in all other cases.